document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -82,6 +83,42 @@ public boolean upsert(Key key, Document document) throws IOException {
     }
   }
 
+  /**
+   * Same as existing upsert method, however, extends the support with condition filter and optional
+   * parameter for explicitly controlling insert and update.
+   */
+  @Override
+  public boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
+      throws IOException {
+    StringBuilder upsertQueryBuilder =
+        new StringBuilder(getUpsertSQL(isUpsert != null ? isUpsert : true));
+
+    String jsonString = prepareUpsertDocument(key, document);
+    Params.Builder paramsBuilder = Params.newBuilder();
+    paramsBuilder.addObjectParam(key.toString());
+    paramsBuilder.addObjectParam(jsonString);
+    paramsBuilder.addObjectParam(jsonString);
+
+    if (condition != null && isUpsert) {
+      String conditionQuery = parseFilter(condition, paramsBuilder, "d");
+      if (conditionQuery != null) {
+        upsertQueryBuilder.append(" WHERE ").append(conditionQuery);
+      }
+    }
+
+    try (PreparedStatement preparedStatement =
+        buildPreparedStatement(upsertQueryBuilder.toString(), paramsBuilder.build())) {
+      int result = preparedStatement.executeUpdate();
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: {}", result);
+      }
+      return result > 0;
```

Comment:
what exactly is the result? is it the number of modified records or the id of the update? maybe a more meaningful var name would help?